src/doc/src/reference/resolver.md

```diff
@@ -7,8 +7,10 @@ result of the resolution is stored in the `Cargo.lock` file which "locks" the
 dependencies to specific versions, and keeps them fixed over time.
 
 The resolver attempts to unify common dependencies while considering possibly
-conflicting requirements. The sections below provide some details on how these
-constraints are handled, and how to work with the resolver.
+conflicting requirements. Dependency resolution is, however, an NP-hard problem,
```

Comment:
Nit: I don't think it's the `NP-hard`nes of the problem that causes counterintuitive results. The heuristic comes in in deciding which solution, among the seemingly infinite, valid solutions to pick.

Reply:
I'd argue that they're strongly related if not strictly causative. Or, to put it another way, if the problem *wasn't* NP-hard, I suspect that resolution would produce counterintuitive results much more rarely, possibly never — but also it doesn't feel very helpful to speculate about a hypothetical different problem!

Reply:
Having thought about your comment for three days, and sorry for the slow response, I think you are technically correct. One of the definitions of NP-hard is "exponential to find a solution, polynomial to check a solution". And what want for our example is a problem where "the solutions are so simple that people don't think about the algorithm" but where "there are nearly infinite number of possible solutions". So yes, these seem to be isomorphic definitions.

Which still leaves us with question about how we best communicate with a Rust user. Who is probably reading this section because they are frustrated that cargo gave them the "wrong" solution to a simple problem. (Where "wrong" can be "why did you bloat my binary with more than one version" or "why did you give me this ancient version just because it's somewhere else in my dependency tree".) This user may very well never have heard of NP-hard, I certainly had not when I started using Rust. Even if they have, it may not be a clear connection, as I demonstrated by taking three days to understand your point.

Reply:
I've rephrased this bit (in response to @ehuss's comments as well). I do think that it would be illuminating if we can find a good way to introduce the idea that dependency resolution is not as easy as one might naively think.